### PR TITLE
Introducing UserInputMapper

### PR DESCRIPTION
**Pull request by samcake**
- The class UserInputMapper is in charge of bindings keys and any device inputs identified with a unique Input to the actual few "Actions"

- Replace all the code happening in Application::key*Event to control the _myAvatar driveKeys, instead it is happening in the Application::update right after the userINputMapper has been updated

- For now the Actions of the UserInputMapper are hard coded, we ll need to change that soon and make the actions created at runtime

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: b7bacccce770da50e88297581eec8b9ac22dedd4 available at: https://gist.github.com/fdb9589dbd3f662df529<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3921156978061ad3ed71de84403cab4206eeccc3 available at: https://gist.github.com/fdb9589dbd3f662df529<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.dmg

File: interface/src/ui/UserInputMapper.h [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    QPoint _lastCursor;
+    glm::vec2 _lastTouch;
+    bool _isTouching = false;
+    
+    glm::vec2 evalAverageTouchPoints(const QList<QTouchEvent::TouchPoint>& points) const;
+    std::chrono::high_resolution_clock _clock;
+    std::chrono::high_resolution_clock::time_point _lastTouchTime;
+};
+
+#endif // hifi_UserInputMapper_h
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.h]: put this in a sep file

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -1483,7 +1414,10 @@ void Application::mousePressEvent(QMouseEvent* event, unsigned int deviceID) {
 
 
     if (activeWindow() == _window) {
-        if (event->button() == Qt::LeftButton) {
+
+        _keyboardMouseDevice.mousePressEvent(event);
```
Comment (birarda) [inline interface/src/Application.cpp]: indentation

File: interface/src/ui/UserInputMapper.cpp [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    // For touch event, we need to check that the last event is not too long ago
+    // Maybe it's a Qt issue, but the touch event sequence (begin, update, end) is not always called properly
+    // The following is a workaround to detect that the touch sequence is over in case we didn;t see the end event
+    if (_isTouching) {
+        const auto TOUCH_EVENT_MAXIMUM_WAIT = 100; //ms
+        auto currentTime = _clock.now();
+        auto sinceLastTouch = std::chrono::duration_cast<std::chrono::milliseconds>(currentTime - _lastTouchTime);
+        if (sinceLastTouch.count() > TOUCH_EVENT_MAXIMUM_WAIT) {
+            _isTouching = false;
+        }
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: can we remove this hack from look.js if it's here?
Or is look.js deprecated all together anyways.

File: interface/src/ui/UserInputMapper.cpp [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    return averagePoint;
+}
+
+
+void KeyboardMouseDevice::touchBeginEvent(const QTouchEvent* event) {
+    _isTouching = event->touchPointStates().testFlag(Qt::TouchPointPressed);
+    _lastTouch = evalAverageTouchPoints(event->touchPoints());
+    _lastTouchTime = _clock.now();
+}
+void KeyboardMouseDevice::touchEndEvent(const QTouchEvent* event) {
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    _isTouching = false;
+    _lastTouch = evalAverageTouchPoints(event->touchPoints());
+    _lastTouchTime = _clock.now();
+}
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        _axisStateMap[makeInput(TOUCH_AXIS_Y_NEG).getChannel()] = (currentMove.y > 0 ? currentMove.y : 0.0f);
+    }
+
+    _lastTouch = currentPos;
+}
+
+UserInputMapper::Input KeyboardMouseDevice::makeInput(Qt::Key code) {
+    return UserInputMapper::Input(_deviceID, code & KEYBOARD_MASK, UserInputMapper::ChannelType::BUTTON);
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(Qt::MouseButton code) {
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+            return UserInputMapper::Input();
+    };
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(KeyboardMouseDevice::MouseAxisChannel axis) {
+    return UserInputMapper::Input(_deviceID, axis, UserInputMapper::ChannelType::AXIS);
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(KeyboardMouseDevice::TouchAxisChannel axis) {
+    return UserInputMapper::Input(_deviceID, axis, UserInputMapper::ChannelType::AXIS);
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(KeyboardMouseDevice::TouchButtonChannel button) {
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+        case Qt::MiddleButton:
+            return UserInputMapper::Input(_deviceID, MOUSE_BUTTON_MIDDLE, UserInputMapper::ChannelType::BUTTON);
+        default:
+            return UserInputMapper::Input();
+    };
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(KeyboardMouseDevice::MouseAxisChannel axis) {
+    return UserInputMapper::Input(_deviceID, axis, UserInputMapper::ChannelType::AXIS);
+}
+UserInputMapper::Input KeyboardMouseDevice::makeInput(KeyboardMouseDevice::TouchAxisChannel axis) {
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+
+UserInputMapper::DeviceProxy::Pointer UserInputMapper::getDeviceProxy(const Input& input) {
+    auto device = _registeredDevices.find(input.getDevice());
+    if (device != _registeredDevices.end()) {
+        return (device->second);
+    } else {
+        return DeviceProxy::Pointer();
+    }
+}
+bool UserInputMapper::addInputChannel(Action action, const Input& input, float scale) {
```
Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: space here

File: interface/src/ui/UserInputMapper.cpp (hunk 3 above)
Comment (samcake) [inline interface/src/ui/UserInputMapper.cpp]: YEp, i got that from look.js and yes this code should actually deprecate look.js

Comment (birarda) [inline interface/src/ui/UserInputMapper.cpp]: If look.js is no longer required can we remove alltogether or remove from default scripts?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 936eefd4f0f1b6422d36cd50d614b9ecf1c7241f available at: https://gist.github.com/6815f053237396cb668d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c7153bcc14b69e67d0b5c0ee91f1e13e952d5643 available at: https://gist.github.com/fdb9589dbd3f662df529<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fad623ae48c2da5a9d9098dde925480c03658f1e available at: https://gist.github.com/6815f053237396cb668d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4753.dmg

**Pull request merged**